risc0/zkvm/src/host/server/session.rs

```diff
@@ -171,16 +180,22 @@ impl Session {
         // Construct the Output struct for the session, checking internal consistency.
         // NOTE: The Session output is distinct from the final Segment output because in the
         // Session output any proven assumptions are not included.
+        self.claim_with_assumptions(self.assumptions.iter().map(|(_, x)| x))
+    }
+
+    pub(crate) fn claim_with_assumptions<'a>(
```

Comment:
Instead of adding this method, it should be possible (and semantically consistent) to add a `filter_map` below that removed assumptions that have a `pending_zkr`. The framing here being that `session.claim()` returns the claim that the caller can "expect to prove" with this session object. I.e. `session.claim == prover.prover_session(&ctx, &session)?.receipt.claim()`. The code in `ProverImpl::prove_session` and `CompositeReceipt::claim` handles this, and if we filter out assumptions that have a pending ZKR, then the results from `Session::claim` should match.

Reply:
I'm not sure follow. I think we can revisit this as we iterate on this. I wouldn't call this integration done for a few more PRs.

Reply:
Sounds good, and can definitely address it later.

The core idea of this suggestion is that each `pending_zkr` represents a receipt that will be generated as part of proving the execution (e.g. using `prove_session` or using Bonsai's workflow implementation). When we generate the claim for the session, we can assume that will happen, and so drop any assumptions that will be proven by a `pending_zkr`.

There is a tricky bit about determining whether or not a given `ProveZkrRequest` resolves a particular assumption, because assumptions encode the `control_root` and ZKR requests the `control_id`. If we make the assumption that every `ProveZkrRequest` is proven against the singleton set of control IDs (which is all we handle right now and is assumed elsewhere to be the case), then this linking can happen simply be constructing the `control_root` for the singleton set `[control_id]`.